### PR TITLE
chapter_9: Advanced login

### PR DESCRIPTION
**Pull request by ngyihkai**
Ready

### CODE REVIEW

**Comment by ngyihkai** [comment]
Ready

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -3,10 +3,11 @@ def new
   end
 
   def create
-    user = User.find_by email: params[:session][:email].downcase
-    if user&.authenticate params[:session][:password]
-      log_in user
-      redirect_to user
+  @user = User.find_by(email: params[:session][:email].downcase)
```
Comment (huongnt-2545) [inline app/controllers/sessions_controller.rb]: indent

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -4,7 +4,15 @@ def logged_in(user)
   end
 
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by id: user_id
+    elsif (user_id = cookies.signed[:user_id])
+      @user = User.find_by id: user_id
+    if @user && @user.authenticated?(cookies[:remember_token])
```
Comment (huongnt-2545) [inline app/helpers/sessions_helper.rb]: indent
```suggestion
      if @user && @user.authenticated?(cookies[:remember_token])
```

File: app/models/user.rb [hunk 3]
```diff
@@ -1,24 +1,39 @@
...
   validates :name, presence: true
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
   validates :email, presence: true, length: {maximum: 255}, format: {with: VALID_EMAIL_REGEX}, uniqueness: true
 
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
+  end
+
+  has_secure_password
```
Comment (huongnt-2545) [inline app/models/user.rb]: move this line place after validate scope
```
validates :email, presence: true, length: {maximum: 255}, format: {with: VALID_EMAIL_REGEX}, uniqueness: true
has_secure_password

# something method
```

**Comment by ngyihkai** [comment]
Ready

File: app/models/user.rb [hunk 4]
```diff
@@ -1,24 +1,38 @@
...
-           else
-            BCrypt::Engine.cost
-           end
-    BCrypt::Password.create string, cost: cost
+  class << self
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+
+    def User.digest string
```
Comment (ghost) [inline app/models/user.rb]:
```suggestion
    def digest string
```

**Review by ghost — COMMENTED** [review]
cmt

File: app/models/user.rb [hunk 5]
```diff
@@ -1,24 +1,38 @@
...
+      cost = if ActiveModel::SecurePassword.min_cost
+              BCrypt::Engine::MIN_COST
+             else
+              BCrypt::Engine.cost
+             end
+              BCrypt::Password.create string, cost: cost
+    end
+  end
+
+  def forget
```
Comment (ghost) [inline app/models/user.rb]: Can you log out ???

Comment (ngyihkai) [reply app/models/user.rb]: cannot. is it move it out from the class<<self?

**Comment by ngyihkai** [comment]
Ready

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -4,15 +4,40 @@ def logged_in(user)
...
+
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  def authenticated? remember_token
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
```
Comment (thaodtd-0665) [inline app/helpers/sessions_helper.rb]: this method belongs to model user, you need move it to app/models/user.rb

File: app/models/user.rb (hunk 3 above)
Comment (thaodtd-0665) [reply app/models/user.rb]: You need to fix comment

**Comment by ngyihkai** [comment]
Ready

**Review by thaodtd-0665 — APPROVED** [review]

**Comment by thaodtd-0665** [comment]
merged

**Pull request merged**